### ISSUE TITLE
refactor subdir creation to just use model name not the _est_00x convention

### ISSUE THREAD

**Issue by dpastoor** [metrumresearchgroup/babylon#85]

**Comment by shairozan** [comment]
So I originally just removed the "_est_" from the function which looked ok, until I started thinking about models with numeric indicators. 

For example, if your mod is mod001.mod, what you would end up with are directories like `mod00101` and `mod00102`, so I added a "_" back into the name structure. This makes sure you have a clean delineation from the model name and the run number. 

IE `mod001_01` and `mod001_02`

**Comment by shairozan** [comment]
Took some time to figure out I should be using the mod files as opposed to the LST files. Even though the LST contains parts of the control stream, the mod is purely the instructions, and using the LST will generate a nonmem error about the input and output files being the same. Slow progress :-D

**Comment by shairozan** [comment]
[![Alt text](https://img.youtube.com/vi/9arySYYweu4/0.jpg)](https://www.youtube.com/watch?v=9arySYYweu4)

**Comment by shairozan** [comment]
![ezgif-5-c3c3255768a4](https://user-images.githubusercontent.com/952951/69568755-38bc7600-0f8a-11ea-87b2-1014570aab6f.gif)

**Comment by shairozan** [comment]
![Screen-Recording-2019-11-25-at-1 40 22-PM](https://user-images.githubusercontent.com/952951/69568870-6ef9f580-0f8a-11ea-9c45-de28407d2249.gif)

**Comment by dpastoor** [comment]
Nice work, we can actually simplify this further. So the point of dropping the underscore convention was it became unnecessarily hard to derive where to find the correct latest results. Instead we should just have a dir with the exact name with appropriate flags control behavior around what to do override vs error if already present.

I'm thinking on what that default behavior should be

**Comment by shairozan** [comment]
Ok I think I get what you mean. We basically allow the users to define what based on configuration how to name the folders with a default behavior. It makes sense to use the _ as it's an uncommon character, but you can use whatever the runDelimiter is to split the string and locate the last job.

**Comment by dpastoor** [comment]
actually I mean no delimiter at all - just completely simplify the numeric counters for run names period.

So given a model `acop.mod` --> `acop/` no `acop_01` or anything. That way, it is very simple to establish, given a model name, what the definitive directory where final results are.

This dramatically simplifies any logic around reading/summarizing etc as we know for sure what the folder name would be. It does instead introduce a separate bit of nuance - what are the appropriate behaviors given that folder already exists. 

In general there are 4 scenarios that could happen:

1) the model was previous run incorrectly, and is is being submitted again with a fix

2) the model was run correctly, but results were not sufficient enough to want to keep the results, and instead the model was likewise tweaked to re-run again

3) the model was submitted again accidentally (eg in a loop), when in reality there was no reason to re-run.

4) the model was re-run to attempt to reproduce the results 

In scenario's 1/2 - the desired outcome is to purge the old results, and then create a new directory. Importantly, it should be removed, not just overwritten in case any artifacts from a previous run would not be overwritten (eg during update to a model file, drop one of the output tables, it would keep the old output table version)

In scenario 3, the ideal scenario is a no op with a message to the user why, and if they are actually looking to achieve scenario 4 (or 2), they can suitably clear out and create a new dir.


Here is how I think that can be achieved:

## name flexibility

There should be a flag for the `output_dir`, and given such, the dirname would be the model file name (eg acop.mod --> `acop/`) or 1.ctl --> `1/`. However a user could create an output dir, where it is actual a go string template with that can take the param `{{.Name}}` such that they could then construct a templated output_dir

```
bbi nonmem run 1.mod --output_dir="{{.Name}}_est_01" --> 1_est_01
```

This would then allow multiple models to be expanded properly

```
bbi nonmem run 1.mod 2.mod 3.mod --output_dir="{{.Name}}_01"
```

would drop 1.mod --> `1_01/`, 2.mod --> `1_01` etc

## logic around what to do if output_dir already exists

To fill out later

**Comment by shairozan** [comment]
The only question I really have is how do we know the users intent on submission? Ie how do we know that it was "submitted accidentally, eg in a loop?" I'm not super familiar with this process, just trying to figure out how we would make that decision

Also, when it comes to the concept of knowing what was done and what was the definitive run for a model, wouldn't it make sense for Babylon to store the results of the run / or even all runs in the directory itself?

```yaml
input: "run acop[001:003].mod"
input_directory: "/some/user/path"
options:
  output_template: '{.Name}'
models : 
 - acop001 :
    model_file: "acop001.mod"
    model_hash: "ba970d2ad873a872cb83b8e1f04161b3"
    output_dir: "/some/user/path/acop001"
    errored: false
    completed: "2019-11-26 11:23:14"
 - acop002 :
    model_file: "acop002.mod"
    model_hash: "efa963f70d2e2caa6758bd908b99fa40"
    output_dir: "/some/user/path/acop002"
    errored: false
    completed: "2019-11-26 11:23:17"
 - acop003 :
    model_file: "acop003.mod"
    model_hash: "18b59e21527273fdd70facac745afed3"
    output_dir: "/some/user/path/acop003"
    errored: false
    completed: "2019-11-26 11:23:19"
```

Then you could have a `babylon rerun acop0003.mod` or `bablyon run --again acop003.mod` scenario where we would just locate the directory specified in the last run. It can also store all the options provided, or even the raw arguments provided. It could even just locate the map in the file on change and update it. 

Using this approach we could even store MD5 or SHA hashes of the files before the run, which can be used on sequential runs to determine if attempts to run the model without changing the model itself. Similar approach could be done by extracting details of the data file.

**Comment by shairozan** [comment]
Had to mull on it most of yesterday and came to realize what you meant. I was thinking too much in the sense of a run number, but all we care about are models being executed. People were creating naming conventions around the models such that they could control run execution etc.

What I've got now is that the runDir parameter has been replaced with an outputDir parameter which is a golang string template. Defaults to {{ .Name }}. If it's provided, we'll parse it, or parse the default. If a non value is provided we'll bubble up the error.

There's also an overwrite parameter which defaults to true currently. When we process the template, if the output directory exists, we check the contents of the runsettings to see if we should overwrite. If we're set to overwrite, currently it just deletes the directory and moves along (Need to add the mv and roll back logic)

If it is set to not overwrite, however, we send an error back up to user indicating a mismatch of configuration and data state.
